plugins/net_plugin/net_plugin.cpp

```diff
@@ -112,6 +112,145 @@ namespace eosio {
       >
       > peer_block_state_index;
 
+   class address_manager {
+   private:
+       std::vector<peer_address> addresses;
+       mutable std::mutex addresses_mutex;
+
+   public:
+       // time_out for future address time out counting
+       explicit address_manager( uint32_t time_out ){};
+
+       void add_address(const peer_address& address);
+       void add_address_str(const std::string& address, bool is_manual);
+       void add_addresses(const std::vector<std::string>& new_addresses_str, bool is_manual);
+       void remove_address(const peer_address& address);
+       void remove_addresses(const std::vector<peer_address>& addresses_to_remove);
+       void update_address(const peer_address& updated_address);
+       std::vector<peer_address> get_addresses() const;
+       std::vector<peer_address> get_manual_addresses() const;
+       std::vector<peer_address> get_diff_addresses(const std::vector<string>& addresses_exist) const;
+       std::vector<peer_address> get_addresses(uint32_t start, uint32_t count) const;
+       std::vector<std::string> get_addresses_str() const;
+
+       bool has_address(const std::string& address_str) const;
+
+   };
+
+
+    void address_manager::add_address(const peer_address& address) {
+        std::unique_lock<std::mutex> lock(addresses_mutex);
+        // if same address, skip it, ignore other properties
+        if(std::find(addresses.begin(), addresses.end(), address) == addresses.end()) {
+            fc_dlog( logger, "Address Manager add_address: ${host} ${port} ${type}", ("host",address.host)("port",address.port)("type", address_type_str(address.address_type)) );
+            addresses.push_back(address);
+        }
+    }
+
+    void address_manager::add_address_str(const std::string& address, bool is_manual ) {
+        peer_address addr = peer_address::from_str(address, is_manual);
+        // same address with different configurations is ignored
+        add_address(addr);
+    }
+
+    void address_manager::add_addresses(const std::vector<std::string>& new_addresses, bool is_manual ) {
+        std::unique_lock<std::mutex> lock(addresses_mutex);
```

Comment:
In my opinion, because we will add timeout or connectivity checks to the address pool in the future, frequent modifications to the addresses vector by multiple threads may cause issues if another thread deletes an element from the vector while an iterator is being incremented. this modification may result in undefined behavior and locking the entire loop may be necessary.

Reply:
If you put mutex lock just before if statement: `if (std::find(addresses.begin(), addresses.end(), addr) == addresses.end())` should be fine since the `addresses` vector will be locked from further modification.

However, as previously mentioned, you'd have to simulate different situations and measure whether there is any real benefit to it. Since leap is designed to be highly performant, it's reasonable to assume that every unlocked mutex would result in improved throughput. Than again maybe it's not the case here.

Reply:
https://github.com/AntelopeIO/leap/commit/9b213992192e0c09c7f579ddff42a504f756af24

In the given scenario, with 1000 threads concurrently adding 1000 new addresses to the existing 30000 addresses, the first function (i.e., add_addresses) performs better. 

Further testing is needed for more scenarios and both add and remove operations are present.

Reply:
https://github.com/AntelopeIO/leap/pull/918/commits/3567cebddae04441f1dbf85d3e92258c918084bc


Since we are discussing performance, why use vector...

Code has been updated to use unordered_maps, which better meets the requirements for address management.

Reply:
vector add:
30000, 1000 threads add 1000 address:
<img width="917" alt="image" src="https://user-images.githubusercontent.com/9125363/230210970-159e11cb-0105-4375-942d-bf1b5269e3a9.png">

unordered_maps: 
30000, 1000 threads add 10 address, 300 threads remove 10 address, first is better.
<img width="797" alt="image" src="https://user-images.githubusercontent.com/9125363/230213259-665b7548-aff1-4000-9418-7d354c5aed5c.png">

unordered_maps:
30000, 1000 threads add 1000 address, 300 threads remove 1000 address, result is complex.
<img width="865" alt="image" src="https://user-images.githubusercontent.com/9125363/230214097-24ec15a0-7894-449e-8192-76abbb8a3e18.png">

<img width="839" alt="image" src="https://user-images.githubusercontent.com/9125363/230214158-2c21ec65-051d-468f-982d-e726c5d1e68d.png">